src/sbvr-api/sbvr-utils.coffee

```diff
@@ -353,18 +358,59 @@ getHooks = do ->
 	_getHooks = (request) ->
 		if request.resourceName?
 			resourceName = resolveSynonym(request)
-		getMethodHooks(request.method, request.vocabulary, resourceName)
+		instantiateHooks(
+			getMethodHooks(
+				request.method
+				request.vocabulary
+				resourceName
+			)
+		)
 	_getHooks.clear = -> getMethodHooks.clear()
 	return _getHooks
 
-runHook = (hookName, args) ->
+class Hook
+	constructor: (hookFn) ->
+		@hookFn = hookFn
+		@executed = false
+
+	run: (args...) ->
+		Promise.try =>
+			@hookFn(args...)
+		.finally =>
+			@executed = true
+
+class SideEffectHook extends Hook
+	constructor: (hookFn) ->
+		@rollbackFns = []
+		@rolledBack = false
+		super(hookFn)
+
+	registerRollback: (fn) ->
+		runFn = Promise.method(fn)
```

Comment:
There's no need to wrap the `fn` in a `Promise.method` for `@rollbackFns` because the `Promise.each` in `rollback` will automatically handle the wrapping.  As for the direct call version it probably makes more sense to use `Promise.try(fn)` - can you check if `fn` will be run synchronously in those cases though, as we should make sure it's done async to have the same behaviour whether we've already rolled back or not